libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,333 @@
+                            replace {P = \x => x + 0 = n} (multCommutative q (S d)) prf
+                    in
+                    ItIsFactor $ CofactorExists q nok prf
+
+                (FS pr) =>
+                    ItIsNotFactor $ ProperRemExists q pr nok (
+                            rewrite multCommutative d q in
+                            rewrite sym $ multRightSuccPlus q d in
+                            prf
+                        )
+
+||| For all p greater than 1, if p is a factor of n, then it is NOT a factor
+||| of (n + 1).
+factNotSuccFact : {n, p : Nat} -> GT p 1 -> Factor n p -> NotFactor (S n) p
+factNotSuccFact {n} {p = Z} pGt1 (CofactorExists q positN prf) =
+        absurd $ succNotLTEzero pGt1
+factNotSuccFact {n} {p = S Z} pGt1 (CofactorExists q positN prf) =
+        absurd . succNotLTEzero $ fromLteSucc pGt1
+factNotSuccFact {n} {p = S (S k)} pGt1 (CofactorExists q positN prf) =
+        let r = FZ in -- remember it's remainders precedessor
+        ProperRemExists q r (lteSuccRight positN) (
+            rewrite prf in
+            rewrite plusCommutative n 1 in
+            Refl
+        )
+
+||| 1 is a common factor of any pair of natural numbers.
+oneCommonFactor : (a, b : Nat) -> {auto aok : LTE 1 a} -> {auto bok : LTE 1 b} -> CommonFactor a b 1
+oneCommonFactor a b {aok} {bok} = CommonFactorExists 1
+        (CofactorExists a aok (rewrite plusZeroRightNeutral a in Refl))
+        (CofactorExists b bok (rewrite plusZeroRightNeutral b in Refl))
+
+||| Any natural number is a common factor of itself and itself.
+selfIsCommonFactor : (a : Nat) -> {auto ok : LTE 1 a} -> CommonFactor a a a
+selfIsCommonFactor Z {ok} = absurd $ succNotLTEzero ok
+selfIsCommonFactor (S k) = CommonFactorExists (S k) (selfFactor $ S k) (selfFactor $ S k)
+
+-- Some helper definitions only for internal use of gcd procedure.
+namespace GCD
+    %access private
+
+    data Search : Type where
+        SearchArgs : (a, b : Nat) -> LTE b a -> {auto bNonZero : LTE 1 b} -> Search
+
+    left : Search -> Nat
+    left (SearchArgs l _ _) = l
+
+    right : Search -> Nat
+    right (SearchArgs _ r _) = r
+
+    Sized Search where
+        size (SearchArgs a b _) = a + b
+
+    step : (x : Search) ->
+        (rec : (y : Search) -> Smaller y x ->  (f : Nat ** GCD (left y) (right y) f)) ->
+        (f : Nat ** GCD (left x) (right x) f)
+    step (SearchArgs Z _ bLteA {bNonZero}) _ = absurd . succNotLTEzero $ lteTransitive bNonZero bLteA
+    step (SearchArgs _ Z _ {bNonZero}) _ = absurd $ succNotLTEzero bNonZero
+    step (SearchArgs (S a) (S b) bLteA {bNonZero}) rec with (divMod (S a) (S b))
+        | Fraction (S a) (S b) q FZ prf =
+            let sbIsFactor = the (plus q (mult b q) = S a) $
+                    rewrite multCommutative b q in
+                    rewrite sym $ multRightSuccPlus q b in
+                    replace {P = \x => x = S a} (plusZeroRightNeutral (q * S b)) prf
```

Comment:
`replace` can be replaced with

```idris
                    rewrite sym prf in
                    sym $ plusZeroRightNeutral (q * S b)
```

IME explicit use of `replace` is a "proof smell", and there is almost always a nicer way to do it.

Reply:
I try to use `rewrite` rather than `replace` whenever I can, but it's not always possible. There is the huge difference between them, however. `replace` makes a substitution in a proof you already have to create another proof, whereas `rewrite` makes a substitution in the goal (which is a proof you do *not* have yet) instead. In practice it means that with `rewrite` you work kind of backwards, rewriting your goal until it matches something you already have, while `replace` has a more direct approach. Also with `rewrite` you can only use equality `(=)` proofs (the same with `replace` of course), which means that you you need use a custom proof-altering function like `cong` or something in the middle of replacements, you actually can no longer use `rewrite` from that point on. Unless of course I am missing something here.